File: cs/MemRequest.cs
namespace Nijikawa
{
    enum MemRequestType
    {
        Read,
        Write,
    }

    class MemRequest
    {
        public MemRequestType Type { get; private set; }
        public ulong Addr { get; private set; }
        private ResponseReceiverDel ResponseReceiver { get; set; }

        public delegate void ResponseReceiverDel(long cycle,
                MemResponse response);

        public MemRequest(MemRequestType type, ulong addr,
                ResponseReceiverDel responseReceiver)
        {
            Type = type;
            Addr = addr;
            ResponseReceiver = responseReceiver;
        }

        public void Respond(long cycle) {
            if (ResponseReceiver != null) {
                ResponseReceiver(cycle, new MemResponse(Addr));
            }
        }
    }

    class MemResponse
    {
        public ulong Addr { get; private set; }

        public MemResponse(ulong addr)
        {
            this.Addr = addr;
        }
    }

    interface IMemRequestReceiver
    {
        void ReceiveMemRequest(MemRequest req);
    }
}

File: rust/mem_request.rs
use std::u64;

pub enum MemRequestType {
    Read,
    Write,
}

pub struct MemRequest {
    addr: u64,
    req_type: MemRequestType,
    response_receiver: Option<@mut MemResponseReceiver>,
}

pub struct MemResponse {
    addr: u64,
}

pub trait MemRequestReceiver {
    fn receive_request(&mut self, request: ~MemRequest);
}

pub trait MemResponseReceiver {
    fn receive_response(&mut self, cycle: i64, response: ~MemResponse);
}

impl MemRequest {
    pub fn respond(&self, cycle: i64) {
        match self.response_receiver {
            Some(receiver) => receiver.receive_response(
                    cycle, ~MemResponse { addr: self.addr }),
            None => ()
        }
    }
}
